dev-tools/harness.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * Random Simulation harness for testing and benchmarking purposes.
+ *
+ * Pokemon Showdown - http://pokemonshowdown.com/
+ *
+ * @license MIT
+ */
+
+'use strict';
+
+const child_process = require('child_process');
+const shell = cmd => child_process.execSync(cmd, {stdio: 'inherit', cwd: __dirname});
+// `require('../build')` is not safe because `replace` is async.
+shell('node ../build');
+
+const BattleStreams = require('../.sim-dist/battle-stream');
+const Dex = require('../.sim-dist/dex');
+const PRNG = require('../.sim-dist/prng').PRNG;
+const RandomPlayerAI = require('../.sim-dist/examples/random-player-ai').RandomPlayerAI;
+
+const DEFAULT_SEED = [0x09917, 0x06924, 0x0e1c8, 0x06af0];
+const AI_OPTIONS = {move: 0.7, mega: 0.6};
+
+const FORMATS = [
+	'gen7randombattle', 'gen7randomdoublesbattle', 'gen7battlefactory', 'gen6randombattle', 'gen6battlefactory',
+	'gen5randombattle', 'gen4randombattle', 'gen3randombattle', 'gen2randombattle', 'gen1randombattle',
+];
+
+class Runner {
+	constructor(options) {
+		this.prng = new PRNG(options.seed);
+		this.format = options.format;
+		this.totalGames = options.totalGames;
+		this.all = !!options.all;
+		this.logs = !!options.logs;
+		this.silent = !this.logs || !!options.silent;
+		this.async = !!options.async;
+		this.p1options = options.p1options;
+		this.p2options = options.p2options;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let format, lastFormat;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				await Promise.all(games);
+				games = [];
+			}
+
+			let seed = this.prng.seed;
+			try {
+				const game = this.runGame(format);
+				if (!this.async) await game;
+				games.push(game);
+			} catch (e) {
+				console.error(`Run \`node dev-tools/harness 1 --format=` +
+					`${format} --seed=${seed.join(',')} --logs\` to debug:\n`, e);
+			}
+			lastFormat = format;
+		}
+
+		// Calculate how many games failed (failures weren't added to `games`).
+		return this.totalGames - (await Promise.all(games)).length;
+	}
+
+	async runGame(format) {
+		const streams = BattleStreams.getPlayerStreams(new BattleStreams.BattleStream());
+		const spec = {formatid: format, seed: this.prng.seed};
+		const p1spec = {name: "Bot 1", team: this.generateTeam(format)};
+		const p2spec = {name: "Bot 2", team: this.generateTeam(format)};
+
+		/* eslint-disable no-unused-vars */
+		const p1 = new RandomPlayerAI(streams.p1,
+			Object.assign({seed: this.nextSeed()}, this.p1options));
+		const p2 = new RandomPlayerAI(streams.p2,
+			Object.assign({seed: this.nextSeed()}, this.p2options));
```

Comment:
The usage of ``.nextSeed()`` here implies that ``p1`` and ``p2`` issue the same commands off by one turn, right? That makes it seem so unsound...

However, just like my seed comment, I wonder if we should care at all.